### PR TITLE
Add basic lua List module

### PR DESCRIPTION
**Pull request by tarleb**
The List module is automatically loaded, but not assigned to a global
variable. It can be included in filters by calling `List = require
'List'`.

Lists of blocks, lists of inlines, and lists of classes are now given
`List` as a metatable, making working with them more convenient. E.g.,
it is now possible to concatenate lists of inlines using Lua's
concatenation operator `..` (requires at least one of the operants to
have `List` as a metatable):

    function Emph (emph)
      local s = {pandoc.Space(), pandoc.Str 'emphasized'}
      return pandoc.Span(emph.content .. s)
    end

Closes: #4081

### CODE REVIEW

**Comment by tarleb** [comment]
I don't consider this ready to merge yet. I just wanted to have something worth discussing.

**Comment by jgm** [comment]
Is there a reason the module name is capitalized here, but
not in other places (e.g. pandoc, text)?

**Comment by tarleb** [comment]
The module is also a class, which we capitalize in existing code. We could keep the module name lowercase, importing it into an capitalized var, but I felt that capitalizing the module would be the lesser evil.

**Comment by tarleb** [comment]
I opted for functions which (slightly) resemble those in lua's `string` module and JavaScript's Array prototype.  A Common Lisp inspired function name was used where additional flexibility is desired (`find_if`). The library contains only functions which I considered essential, more can be added later.  From my point of view, this is ready to merge now.

The module loading code needs more work, but I'm planning to refactor this anyway. I'll address it in a separate commit/request.

File: data/List.lua [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+end
+
+--- Returns a (shallow) copy of the list.
+function List:clone ()
+  local lst = setmetatable({}, getmetatable(self))
+  List.extend(lst, self)
+  return lst
+end
+
+--- Appends the given list to the end of this list.
```
Comment (jgm) [inline data/List.lua]: This comment is wrong, isn't it?

File: data/pandoc.lua [hunk 2]
```diff
@@ -262,7 +264,7 @@ M.DefinitionList = M.Block:create_constructor(
 M.Div = M.Block:create_constructor(
   "Div",
   function(content, attr)
-    return {c = {attr or M.Attr(), content}}
+    return {c = {attr or M.Attr(), List:new(content)}}
```
Comment (jgm) [inline data/pandoc.lua]: I'm curious why this change is needed?

**Comment by jgm** [comment]
I've made a couple small comments in the code.

My main request is: can we also have a `filter` function that returns a new list with the elements of the old list that match a condition?

File: data/List.lua (hunk 1 above)
Comment (tarleb) [reply data/List.lua]: You are right, that's very wrong.

File: data/pandoc.lua (hunk 2 above)
Comment (tarleb) [reply data/pandoc.lua]: It's just for convenience, making it possible to write `div.content:find_if(fn)` rather than `List.find_if(div.content, fn)`. Users won't have to `require 'List'`. It doesn't appear to affect performance in a significant way.

**Comment by tarleb** [comment]
Added `filter` and fixed the comment.  I also renamed the module to `pandoc.List` to avoid future name clashes.

**Comment by johnd0e** [comment]
```
local M = {
  _VERSION = "0.1.0"
}
```
This variable is not used. Intentionally?
Or maybe `_VERSION` should go into `List` table?

And question: where can we find documentation for this module, sample of usage etc?

**Comment by tarleb** [comment]
> Or maybe _VERSION should go into List table?

Thanks, fixed.

There is no generated documentation yet. I guess we could either add it to lua-filters, or provide a link to a ldoc generated file. The former seems the better option, I'll check what needs to be done for this.

**Comment by tarleb** [comment]
This is weird. My last changes, including the filter function, didn't make it to master. I must have pushed to the wrong branch.  I'll fix it.
